src/services/services.ts

```diff
@@ -5288,6 +5294,149 @@ namespace ts {
+                    fileName: entry.fileName
+                });
+            }
+            else if (definitionIsImplementation(node, typeChecker)) {
+                const definitions = getDefinitionAtPosition(fileName, position);
+                forEach(definitions, (definition: DefinitionInfo) => {
+                    entries.push({
+                        textSpan: definition.textSpan,
+                        fileName: definition.fileName
+                    });
+                });
+            }
+            else {
+                // Do a search for all references and filter them down to implementations only
+                const result = getReferencedSymbolsForNode(node, program.getSourceFiles(), /*findInStrings*/false, /*findInComments*/false, /*implementations*/true);
+
+                forEach(result, referencedSymbol => {
+                    if (referencedSymbol.references) {
+                        forEach(referencedSymbol.references, entry => {
+                            entries.push({
+                                textSpan: entry.textSpan,
+                                fileName: entry.fileName
+                            });
+                        });
+                    }
+                });
+            }
+
+            return entries;
+        }
+
+        /**
+         * Returns true if the implementation for this node is the same as its definition
+         */
+        function definitionIsImplementation(node: Node, typeChecker: TypeChecker) {
+            if (node.kind === SyntaxKind.SuperKeyword || node.kind === SyntaxKind.ThisKeyword) {
+                return true;
+            }
+
+            if (node.parent.kind === SyntaxKind.PropertyAccessExpression || node.parent.kind === SyntaxKind.ElementAccessExpression) {
+                const expression = (<ElementAccessExpression | PropertyAccessExpression>node.parent).expression;
+
+                // Members of "this" and "super" only have one possible implementation, so no need to find
+                // all references. Similarly, for the left hand side of the expression it only really
+                // makes sense to return the definition
+                if (node === expression || expression.kind === SyntaxKind.SuperKeyword || expression.kind === SyntaxKind.ThisKeyword) {
+                    return true;
+                }
+
+                // Check to see if this is a property that can have multiple implementations by determining
+                // if the parent is an interface (or class, or union/intersection)
+                const expressionType = typeChecker.getTypeAtLocation(expression);
+                if (expressionType && expressionType.getFlags() & (TypeFlags.Class | TypeFlags.Interface | TypeFlags.UnionOrIntersection)) {
+                    return false;
+                }
+
+                // Also check the right hand side to see if this is a type being accessed on a namespace/module

```

Comment:
This check is to see if this property access expression is a reference to a type in a namespace or module. For example:
```typescript
namespace Foo {
    interface Bar {}
}

// Right hand side is an interface, left hand side is not
Foo.Bar
```